File: Leetcode/129.求根节点到叶节点数字之和.cpp
/*
 * @lc app=leetcode.cn id=129 lang=cpp
 *
 * [129] 求根节点到叶节点数字之和
 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

/*2021-12-06-17-53*/

// https://github.com/fengwei2002/Algorithm
// solution link: 
// https://leetcode-cn.com/problems/sum-root-to-leaf-numbers/solution/lc129-fengwei2002-by-kycu-jr9s/

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// @lc code=start


class Solution {
public:
    int ans = 0;

    int sumNumbers(TreeNode* root) {
        if (root) dfs(root, 0);
        return ans;
    }

    void dfs(TreeNode* root, int number) {
        number = number * 10 + root->val;
        if (!root->left && !root->right) ans += number;
        if (root->left) dfs(root->left, number);
        if (root->right) dfs(root->right, number);
    }
};
// @lc code=end

File: Leetcode/129.求根节点到叶节点数字之和.go
/*
 * @lc app=leetcode.cn id=129 lang=golang
 *
 * [129] 求根节点到叶节点数字之和
 */
package leetcode

import (
	"strconv"
)

/* --- 2021-11-19-14-58 --- */

// https://github.com/fengwei2002/Algorithm
// solution link:
// https://leetcode.cn/problems/sum-root-to-leaf-numbers/solution/lc129-fengwei2002-by-kycu-2anv/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start

var stringArray []string

func sumNumbers(root *TreeNode) int {
	// 计算所有路径生成的数字的 和
	ans := 0
	stringArray = make([]string, 0)

	dfs(root, strconv.Itoa(root.Val))

	for _, v := range stringArray {
		t, _ := strconv.Atoi(v)
		ans += t
	}
	return ans
}

func dfs(root *TreeNode, path string) {
	if root.Left == nil && root.Right == nil {
		stringArray = append(stringArray, path)
		return
	}

	if root.Left != nil {
		path += strconv.Itoa(root.Left.Val)
		dfs(root.Left, path)
		path = path[:len(path)-1]
	}
	if root.Right != nil {
		path += strconv.Itoa(root.Right.Val)
		dfs(root.Right, path)
		path = path[:len(path)-1]
	}

	return
}

// @lc code=end
